File: apps/python/src/exercises/web/http_access_logging.py
"""Skip noisy http.request access lines for routine probe/scrape traffic (log failures only)."""

from __future__ import annotations

QUIET_GET_PATHS = frozenset({"/metrics"})
QUIET_POST_STATUSES: dict[str, frozenset[int]] = {
    "/api/auth/ensure": frozenset({200}),
}


def request_pathname(path: str) -> str:
    pathname = path.split("?", 1)[0]
    if len(pathname) > 1 and pathname.endswith("/"):
        pathname = pathname.rstrip("/")
    return pathname


def should_log_http_access(method: str, path: str, status: int | None = None) -> bool:
    """When False, skip http.request received/completed lines for this request."""
    pathname = request_pathname(path)
    upper = method.upper()

    if upper == "GET" and pathname in QUIET_GET_PATHS:
        if status is None:
            return False
        return status != 200

    quiet_statuses = QUIET_POST_STATUSES.get(pathname)
    if upper == "POST" and quiet_statuses is not None:
        if status is None:
            return False
        return status not in quiet_statuses

    return True

File: apps/java/src/main/java/com/example/demo/observability/HttpAccessLogging.java
package com.example.demo.observability;

import java.util.Map;
import java.util.Set;

/** Skip noisy http.request access lines for routine probe/scrape traffic (log failures only). */
public final class HttpAccessLogging {

  private static final Set<String> QUIET_GET_PATHS =
      Set.of("/actuator/prometheus", "/api/observability/health");

  private static final Map<String, Set<Integer>> QUIET_POST_STATUSES =
      Map.of("/api/auth/ensure", Set.of(200));

  private HttpAccessLogging() {}

  static String requestPathname(String path) {
    if (path == null || path.isEmpty()) {
      return "/";
    }
    int query = path.indexOf('?');
    return query == -1 ? path : path.substring(0, query);
  }

  /** When false, skip both received and completed http.request lines for this request. */
  public static boolean shouldLogHttpAccess(String method, String path, int status) {
    String pathname = requestPathname(path);
    if (method != null && "GET".equalsIgnoreCase(method)) {
      if (QUIET_GET_PATHS.contains(pathname)) {
        return status != 200;
      }
      return true;
    }
    if (method != null && "POST".equalsIgnoreCase(method)) {
      Set<Integer> quietStatuses = QUIET_POST_STATUSES.get(pathname);
      if (quietStatuses != null) {
        return !quietStatuses.contains(status);
      }
    }
    return true;
  }
}
